### PR TITLE
Prohibit append-only writes targeting backing indices directly.

### PR DESCRIPTION
**Pull request by martijnvg**
Append-only writes can only target the corresponding data stream.

Depends on #57721
Relates to #53100

### CODE REVIEW

**Review by henningandersen — COMMENTED** [review]
I left a couple initial comments. I think we should be able to check this without relying on the naming conventions - and without looking up the template.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 1]
```diff
@@ -271,6 +276,44 @@ public void onFailure(Exception e) {
...
     }
 
+    static void prohibitAppendOnlyWritesInBackingIndices(DocWriteRequest<?> writeRequest, Metadata metadata) {
+        if (writeRequest.index().startsWith(".ds-") == false) {
+            // This is definitively not a backing index of data stream, skip further checking
+            return;
+        }
+
+        // Extract data stream name and check if a composable template with a data stream definition would match with it:
+        int indexOfLastDash = writeRequest.index().lastIndexOf('-');
```
[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:286]: The reason I was concerned about using the lookup was, backing indices that don't yet exist. But thinking more about it, I don't I have to be concerned about it. (given that in security should only apply a privilege from a data stream to a backing index if that backing index exists.)

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 2]
```diff
@@ -271,6 +276,44 @@ public void onFailure(Exception e) {
         }
     }
 
+    static void prohibitAppendOnlyWritesInBackingIndices(DocWriteRequest<?> writeRequest, Metadata metadata) {
+        if (writeRequest.index().startsWith(".ds-") == false) {
```
[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:280]: I will make that change 👍

**Comment by martijnvg** [comment]
The rollover doc snippet fails, that tests rolling over a data stream. This doc snippet test reindex in to a backing index of a data stream and this is no longer allowed with this change. Changing the snippet to reindex into the data stream, requires making reindex data stream compatible. I will open another PR for this and then this PR can get merged.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 3]
```diff
@@ -271,6 +275,35 @@ public void onFailure(Exception e) {
...
+            return;
+        }
+        if (indexAbstraction.getType() != IndexAbstraction.Type.CONCRETE_INDEX) {
+            return;
+        }
+        if (indexAbstraction.getParentDataStream() == null) {
+            return;
+        }
+
+        // At this point with write op is targeting a data stream directly,so
```
Comment (danhermann) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:290]: I think at this point it's targeting a backing index of a data stream directly?

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 4]
```diff
@@ -271,6 +275,35 @@ public void onFailure(Exception e) {
...
+        DocWriteRequest.OpType opType = writeRequest.opType();
+        // CREATE op_type is considered append-only and
+        // INDEX op_type is considered append-only when no if_primary_term and if_seq_no is specified.
+        // (the latter maybe an update, but at this stage we can't determine that. In order to determine
+        // that an engine level change is needed and for now this check is sufficient.)
+        if (opType == DocWriteRequest.OpType.CREATE ||
+            (opType == DocWriteRequest.OpType.INDEX && writeRequest.ifPrimaryTerm() == UNASSIGNED_PRIMARY_TERM &&
+                writeRequest.ifSeqNo() == UNASSIGNED_SEQ_NO)) {
+            throw new IllegalArgumentException("append-only write targeting backing indices is disallowed," +
+                "target corresponding data stream instead");
```
Comment (danhermann) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:303]: Minor: added space in error message
```suggestion
            throw new IllegalArgumentException("append-only write targeting backing indices is disallowed, " +
                "target corresponding data stream instead");
```

**Review by danhermann — APPROVED** [review]
LGTM. Left a couple minor comments below

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 5]
```diff
@@ -271,6 +275,35 @@ public void onFailure(Exception e) {
         }
     }
 
+    static void prohibitAppendOnlyWritesInBackingIndices(DocWriteRequest<?> writeRequest, Metadata metadata) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:278]: nit: I think

```suggestion
    static void prohibitAppendWritesInBackingIndices(DocWriteRequest<?> writeRequest, Metadata metadata) {
```

is more precise in that we try to prohibit any direct append into background indices?

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 6]
```diff
@@ -271,6 +275,35 @@ public void onFailure(Exception e) {
...
+
+        DocWriteRequest.OpType opType = writeRequest.opType();
+        // CREATE op_type is considered append-only and
+        // INDEX op_type is considered append-only when no if_primary_term and if_seq_no is specified.
+        // (the latter maybe an update, but at this stage we can't determine that. In order to determine
+        // that an engine level change is needed and for now this check is sufficient.)
+        if (opType == DocWriteRequest.OpType.CREATE ||
+            (opType == DocWriteRequest.OpType.INDEX && writeRequest.ifPrimaryTerm() == UNASSIGNED_PRIMARY_TERM &&
+                writeRequest.ifSeqNo() == UNASSIGNED_SEQ_NO)) {
+            throw new IllegalArgumentException("append-only write targeting backing indices is disallowed, " +
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:302]: "append-only" in this message could cause some confusion. In principle, the user could be doing an index operation that he expects to update an existing document. We just cannot determine that up front without having him be explicit in the request.

I suggest to turn this into two messages, one for the `op_type` create case and one for the `op_type` index case to make this clearer.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 7]
```diff
@@ -310,7 +343,7 @@ static boolean resolvePipelines(final DocWriteRequest<?> originalRequest, final
                 // the index does not exist yet (and this is a valid request), so match index
                 // templates to look for pipelines in either a matching V2 template (which takes
                 // precedence), or if a V2 template does not match, any V1 templates
-                String v2Template = MetadataIndexTemplateService.findV2Template(metadata, indexRequest.index(), false);
+                String v2Template = findV2Template(metadata, indexRequest.index(), false);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:346]: nit: this seems unrelated now.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 8]
```diff
@@ -162,6 +165,7 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
         final Metadata metadata = clusterService.state().getMetadata();
         final Version minNodeVersion = clusterService.state().getNodes().getMinNodeVersion();
         for (DocWriteRequest<?> actionRequest : bulkRequest.requests) {
+            prohibitAppendOnlyWritesInBackingIndices(actionRequest, metadata);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:168]: I think the exception thrown will terminate the entire bulk request. Normally, only errors around the bulk structure itself will cause the entire bulk request to fail. Individual item problems are reported back per item instead. I would think this should be per item too.

I think moving the check into [here in `BulkOperation`](https://github.com/elastic/elasticsearch/blob/4f2f0b5ace82681d9a8f8cd9c457d929f15cd29c/server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java#L457) would fix this.

**Review by henningandersen — COMMENTED** [review]
I left a couple of comments, but the main one is about the granularity of reporting the error. I think this type of error should go back on the item level, since it is not entirely about structure of the request, the error is around doing this to the specific indices.

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.data_stream/10_basic.yml [hunk 9]
```diff
@@ -200,3 +200,59 @@ setup:
...
+          - index:
+              _index: .ds-logs-foobar-000001
+          - foo: bar
+          - create:
+              _index: logs-foobar
+          - foo: bar
+  - match: { errors: true }
+  - match: { items.0.create.status: 400 }
+  - match: { items.0.create.error.type: illegal_argument_exception }
+  - match: { items.0.create.error.reason: "index request with op_type=create targeting backing indices is disallowed, target corresponding data stream instead" }
```
Comment (henningandersen) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.data_stream/10_basic.yml:248]: I wonder if we should help the user and include both the concrete index name and the data stream to target instead?

**Review by henningandersen — APPROVED** [review]
LGTM.
